porch/func/internal/podevaluator.go

```diff
@@ -65,30 +74,109 @@ func NewPodEvaluator(namespace, wrapperServerImage string) (Evaluator, error) {
+	for {
+		time.Sleep(pe.gcScanInternal)
+		err = pe.kubeClient.List(context.Background(), podList, client.InNamespace(pe.namespace))
+		if err != nil {
+			klog.Warningf("unable to list pods in namespace %v: %w", pe.namespace, err)
+			continue
+		}
+		for _, pod := range podList.Items {
+			lastUse, found := pod.Annotations[lastUseTimeAnnotation]
+			// If a pod doesn't have a last-use annotation, we patch it.
+			if !found {
+				if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
+					klog.Warningf("can't patch pod %v/%v with the last-use annotation: %w", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				lu, err := strconv.ParseInt(lastUse, 10, 64)
+				if err != nil {
+					klog.Warningf("unable to convert the Unix time string to int64: %w", err)
+					// If the annotation is ill-formatted, we patch it with the current time and will try to GC it later.
+					if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
+						klog.Warningf("can't patch pod %v/%v with the last-use annotation: %w", pod.Namespace, pod.Name, err)
+					}
+					continue
+				}
+				if time.Unix(lu, 0).Add(pe.podTTL).Before(time.Now()) {
+					// deleting the grpc connection from the pool
+					image := pod.Spec.Containers[0].Image
+					cc, found := pe.grpcConnectionPool[image]
+					if found {
+						klog.Infof("deleting grpc connection %v for %v", cc.Target(), image)
+						delete(pe.grpcConnectionPool, image)
+					}
+					// deleting the old pod
+					klog.Infof("deleting pod %v/%v", pod.Namespace, pod.Name)
+					err = pe.kubeClient.Delete(context.Background(), &pod)
+					if err != nil {
+						klog.Warningf("unable to delete pod %v/%v: %w", pod.Namespace, pod.Name, err)
+						continue
+					}
+				}
+			}
+		}
 	}
-	img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
-	if err != nil {
-		return nil, err
+}
+
+func (pe *podEvaluator) retrieveOrCreatePod(ctx context.Context, image string) (client.ObjectKey, error) {
+	// Try to retrieve the pod first.
+	// Lookup the pod by label to see if there is a pod that can be reused.
+	// Looking it up locally may not work if there are more than one instance of the function runner,
+	// since the pod may be created by one the other instance and the current instance is not aware of it.
+	// TODO: It's possible to set up a Watch in the fn runner namespace, and always try to maintain a up-to-date local cache.
+	podList := &corev1.PodList{}
+	err := pe.kubeClient.List(ctx, podList, client.InNamespace(pe.namespace), client.MatchingLabels(map[string]string{krmFunctionLabel: transformImageName(image)}))
+	if err == nil && len(podList.Items) > 0 {
+		pod := podList.Items[0]
+		klog.Infof("reusing KRM function evaluator pod %q for %q", pod.Name, image)
+		// Update the last-use annotation in the pod
+		if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
```

Comment:
all of these operations will add latency to the function evaluation. We'll want a fast path to fetch gRPC client from a cache and make a call through it, and then secondary slow paths that handle cases when the client or pod don't exist.

As for last use update, we could also do it after the call (outside of the critical request/response path).

Reply:
Made patching asynchronous